File: sdk/lazorkit-sdk/examples/create_wallet.rs
// Example: Creating a LazorKit wallet with Ed25519 authority
//
// This example demonstrates how to:
// 1. Generate a random wallet ID
// 2. Create a wallet with Ed25519 owner
// 3. Derive the config and vault PDAs

use lazorkit_sdk::basic::wallet::LazorWallet;
use lazorkit_sdk::state::AuthorityType;
use solana_sdk::pubkey::Pubkey;
use solana_sdk::signature::Keypair;
use solana_sdk::signer::Signer;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // 1. Generate wallet ID (random 32 bytes)
    let wallet_id = rand::random::<[u8; 32]>();

    // 2. Generate owner keypair
    let owner = Keypair::new();
    let owner_pubkey = owner.pubkey().to_bytes();

    // 3. Build create wallet transaction
    let builder = LazorWallet::create()
        .with_payer(Pubkey::new_unique()) // Replace with actual payer
        .with_id(wallet_id)
        .with_owner_authority_type(AuthorityType::Ed25519)
        .with_owner_authority_key(owner_pubkey.to_vec());

    // 4. Get the PDAs that will be created
    let (config_pda, vault_pda) = builder.get_pdas();

    println!("Creating LazorKit Wallet:");
    println!("  Wallet ID: {}", hex::encode(wallet_id));
    println!("  Config PDA: {}", config_pda);
    println!("  Vault PDA: {}", vault_pda);
    println!("  Owner: {}", owner.pubkey());

    // In a real application, you would:
    // let connection = ...;  // Your Solana connection
    // let tx = builder.build_transaction(&connection).await?;
    // // Sign and send transaction

    Ok(())
}

File: sdk/examples/create-wallet.ts
import {
    createSolanaRpc,
    createKeyPairSignerFromBytes,
    getAddressEncoder,
    generateKeyPairSigner,
    pipe,
    createTransactionMessage,
    setTransactionMessageFeePayerSigner,
    setTransactionMessageLifetimeUsingBlockhash,
    appendTransactionMessageInstruction,
    signTransactionMessageWithSigners,
    sendAndConfirmTransactionFactory,
    createSolanaRpcSubscriptions
} from '@solana/kit';
import {
    createWalletInstruction,
    findConfigPDA,
    findVaultPDA,
    encodeEd25519Authority,
    AuthorityType,
    generateWalletId,
    LAZORKIT_PROGRAM_ID
} from '../src/index.js';

/**
 * Example: Create a LazorKit wallet using Solana Kit
 */
async function main() {
    // Initialize Solana Kit client
    // Use localhost for development
    const rpc = createSolanaRpc('http://127.0.0.1:8899');
    const rpcSubscriptions = createSolanaRpcSubscriptions('ws://127.0.0.1:8900');
    const sendAndConfirmTransaction = sendAndConfirmTransactionFactory({ rpc, rpcSubscriptions });

    // Generate or load a keypair (using random for example)
    const owner = await generateKeyPairSigner();
    console.log('Owner:', owner.address);

    // Request airdrop if on devnet/localhost (this might fail on public devnet due to limits)
    try {
        console.log('Requesting airdrop...');
        // Cast to any for Lamports nominal type
        await rpc.requestAirdrop(owner.address, 1_000_000_000n as any).send();
        // Wait for airdrop confirmation in a real scenario
        await new Promise(r => setTimeout(r, 2000));
    } catch (e) {
        console.warn('Airdrop failed (expected if non-local):', e);
    }

    // Generate wallet ID
    const walletId = generateWalletId();
    console.log('Wallet ID:', Buffer.from(walletId).toString('hex'));

    // Find PDAs
    const configPDA = await findConfigPDA(walletId);
    const vaultPDA = await findVaultPDA(configPDA.address);

    console.log('Config PDA:', configPDA.address);
    console.log('Vault PDA:', vaultPDA.address);

    // Encode owner authority
    const addressEncoder = getAddressEncoder();
    const ownerBytes = addressEncoder.encode(owner.address);
    // Copy to Uint8Array for mutability
    const ownerAuthority = encodeEd25519Authority(new Uint8Array(ownerBytes));

    // Create instruction
    const instruction = createWalletInstruction({
        config: configPDA.address,
        payer: owner.address,
        vault: vaultPDA.address,
        systemProgram: '11111111111111111111111111111111' as any, // System program
        id: walletId,
        bump: configPDA.bump,
        walletBump: vaultPDA.bump,
        ownerAuthorityType: AuthorityType.Ed25519,
        ownerAuthorityData: ownerAuthority,
        programId: LAZORKIT_PROGRAM_ID
    });

    // Send transaction
    console.log('Creating wallet...');

    const { value: latestBlockhash } = await rpc.getLatestBlockhash().send();

    const transactionMessage = pipe(
        createTransactionMessage({ version: 0 }),
        m => setTransactionMessageFeePayerSigner(owner, m),
        m => setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, m),
        m => appendTransactionMessageInstruction(instruction, m)
    );

    const signedTransaction = await signTransactionMessageWithSigners(transactionMessage);

    const signature = await sendAndConfirmTransaction(signedTransaction as any, { commitment: 'confirmed' });

    console.log('✅ Wallet created!');
    console.log('   Signature:', signature);
    console.log('   Config:', configPDA.address);
    console.log('   Vault:', vaultPDA.address);
}

main().catch(console.error);
